design-proposals/backend-storage-block-migration.md

```diff
@@ -0,0 +1,67 @@
+# Overview
+This proposal is about supporting the migration of all VMs that use a persistent feature, such as persistent vTPM.
+
+To make this document as simple as possible, let's consider that storage in Kubernetes can either be filesystem (FS) or
+block mode, and either ReadWriteOnce (RWO) or ReadWriteMany (RWX) access mode, with storage classes supporting any
+number of combinations of both.
+
+As of KubeVirt 1.3, Here is how each of the 4 combinations is supported by the backend-storage for persistent VM state:
+- RWO FS: supported by backend-storage but makes VMs non-migratable
+- RWX FS: preferred combination, supported by backend-storage and compatible with VM migrations
+- RWO block: not supported yet, but PRs exist to implement it in a non-migratable way 
+- RWX block: same as above, with migratability functional but considered unsafe, because it leads to 2 pods (source and target virt-launchers) mounting the same ext3 filesystem concurrently.
+
+This is the best we can do with the current implementation, which involves mounting the same partition on migration source and target.
+This proposal is about switching to a different model which would instead create a new backend-storage PVC on migration,
+allowing all 4 cases above to have proper migration support.
+
+## Motivation/Goals
+- Users want to be able to use any storage class for backend-storage
+- Users want all VMs with persistent features to be potentially migratable
+
+## Definition of Users
+Any VM owner that wants to use persistent features, such as TPM or EFI
+
+## User Stories
+As a user, I want to be able to seamlessly enable persistent features while keeping the migratability of my VMs.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+- When creating a new VM with persistent features, we use either (first match wins):
+  - The storage class specified in the KubeVirt Custom Resource
+  - The storage class marked as the default for virtualization purposes
+  - The storage class marked as the default for the cluster
+- All new persistent storage PVCs will be created with the name `persistent-state-for-<vm_name>-<random_string>` and the new annotation `persistent-state-for-<vm_name>=true`
+- When starting an existing VM with persistent features, we will look for any PVC with the annotation, or any PVC with the legacy name, for backwards compatibility
+- For volume/access modes:
+  - When multiple volume modes are listed in the StorageProfile of the StorageClass, we will default to filesystem (to avoid partitioning complexity)
+  - When multiple access modes are listed for the selected volume mode, we will arbitrarily default to RWO.
+  - When CDI is not present, or the StorageClass has no StorageProfile, or the StorageProfile has no modes, we'll default to RWO FS.
```

Comment:
RWO FS is the lowest common denominator in k8s you will be hard pressed to find a provisioner that does not support that. Not sure we should even consider any other combinations

Reply:
> RWO FS is the lowest common denominator in k8s you will be hard pressed to find a provisioner that does not support that

That's a very important note. How common is RWX FS which we already support with migration?

There may be operational problems with splitting VM persistence between several storage classes (disaster recovery and storage migration may become more complex, virt-default StorageClass) which may keep this proposal useful.

Reply:
> That's a very important note. How common is RWX FS which we already support with migration?

RWX FS is not very popular.  Shared filesystems have been declining in popularity since object storage entered the scene.  Shared filesystems are generally considered complex and slow. For legacy applications. Avoided whenever possible

> There may be operational problems with splitting VM persistence between several storage classes (disaster recovery and storage migration may become more complex, virt-default StorageClass) which may keep this proposal useful.

I'm not following this.  Where's the splitting happening?  My point is that the same storageclass can be used for block and fs.  For example the rook-ceph-block StorageClass can do RWX block and RWO fs.

Reply:
> For example the rook-ceph-block StorageClass can do RWX block and RWO fs.

I didn't know that. Thanks.

So to support TPM on `rook-ceph-block` we have two options: either ask libvirt to support RWX block, or add our own support for RWO file. This proposal should discuss the pros and cons of each.

Reply:
Well this proposal is all about option 2, but I guess we could touch on option 1 in the intro.
A brief pros/cons:
Option 1 requires some work on our side (multi-PVC support) and a lot of work on libvirt's (add support for swtpm block + work with qemu to design and implement EFI block). The main con IMO is the unreasonable ask of libvirt/qemu to support pre-1961 technology (disks without partitions/filesystems).
Option 2 just requires implementing this design proposal, including ensuring there's no race when rebooting during a migration. Seems a lot saner to me, even if it's more work for us.

Reply:
> work with qemu to design and implement EFI block

Seems like a big chunk of work for another team and would ultimately require 2 PVCs for VMs with EFI+TPM, correct?  With filesystem PVCs, TPM and EFI can share a single PVC, right?

Reply:
Exactly yes

Reply:
Fewer PVCs is big PRO for option 2